src/Streamly/Internal/Data/Zipper.hs

```diff
@@ -0,0 +1,380 @@
+-- License     : BSD3
+-- Maintainer  : streamly@composewell.com
+-- Stability   : experimental
+-- Portability : GHC
+--
+-- A special zipper derived from a stream  to facilitate stream parsing.
+--
+-- = Zipper
+--
+-- The zipper is designed to enable moving backward and forward in the stream
+-- efficiently to be able to implement backtracking parsers. The zipper can
+-- buffer data until the parser indicates that it is not needed, we can move
+-- back in the buffered data if the parser needs to backtrack.
+--
+-- = Checkpointing
+--
+-- We have a checkpointing mechanism built into the zipper to handle
+-- backtracking. Whenever we enter a computation that can fail and has an
+-- alternative to run then we add a checkpoint to the zipper. The checkpoint
+-- ensures that we do not release the buffer upto the checkpoint until this
+-- computation succeeds, if the computation fails we can backtrack to focus the
+-- cursor back to the checkpoint position.
+--
+-- = Future Work
+--
+-- We can possibly have a typeclass providing a zipper interface so that we can
+-- derive a zipper from different data sources efficiently. For example,
+--
+-- * stream zipper
+-- * zippers from different types of arrays
+-- * zipper from a tree of arrays for dynamic in-memory storage
+-- * file handle zipper can utilize the seek to go back and forth in the file
+--
+-- A tree of zippers where a lower level zipper is dependent on (references
+-- data in) a higher level zipper could be useful in implementing distributive
+-- parsers.
+--
+-- See https://hackage.haskell.org/package/zipper for an existing type class.
+-- We need a monadic one.
+
+module Streamly.Internal.Data.Zipper
+    (
+      Zipper (..)
+
+    -- * Construction
+    , nil
+    , fromStream
+    , fromList
+
+    -- * Checkpointing
+    , checkpoint
+    , release
+    , restore
+
+    -- * Parsing
+    , parse
+    )
+where
+
+import Control.Exception (assert, Exception(..))
+import Control.Monad.Catch (MonadCatch, try, SomeException)
+-- import GHC.Types (SPEC(..))
+import Prelude hiding (splitAt)
+
+import Streamly.Internal.Data.Stream.StreamK.Type (Stream(..))
+import Streamly.Internal.Data.SVar (defState)
+
+import qualified Streamly.Internal.Data.Parser.ParserD.Types as PR
+import qualified Streamly.Internal.Data.Stream.StreamK as K
+
+-------------------------------------------------------------------------------
+-- Zipper type
+-------------------------------------------------------------------------------
+
+-- | @Zipper checkpoints lefts rights tail@.  The focus is on the first element
+-- of @rights@.  @lefts@ is buffered data on the right of the cursor.  @tail@
+-- is a stream that is used to generate more data if the cursor moves past
+-- @rights@.
+--
+-- @checkpoints@ is a stack of checkpoints. A new checkpoint is created by a
+-- @checkpoint@ operation. A checkpoint consists of a count that tracks how
+-- many elements we have yielded after the checkpoint was taken.  We need this
```

Comment:
"after the checkpoint was taken" == "after the checkpoint was created"?

Reply:
done.